pkg/apb/bind.go

```diff
@@ -37,41 +37,55 @@ func Bind(instance *ServiceInstance,
 	log.Notice(fmt.Sprintf("ServiceInstance.Description: %s", instance.Spec.Description))
 	log.Notice("============================================================")
 
-	executionContext, err := ExecuteApb("bind", instance.Spec, instance.Context, parameters)
-	defer runtime.Provider.DestroySandbox(
-		executionContext.PodName,
-		executionContext.Namespace,
-		executionContext.Targets,
-		clusterConfig.Namespace,
-		clusterConfig.KeepNamespace,
-		clusterConfig.KeepNamespaceOnError,
-	)
-	if err != nil {
-		log.Errorf("Problem executing apb [%s] bind", executionContext.PodName)
-		return executionContext.PodName, nil, err
-	}
-	k8scli, err := clients.Kubernetes()
-	if err != nil {
-		log.Error("Something went wrong getting kubernetes client")
-		return executionContext.PodName, nil, err
-	}
+	go func() {
+		e.start()
+		metrics.ActionStarted("bind")
+		executionContext, err := e.executeApb(
+			"bind", instance.Spec, instance.Context, parameters)
+		defer runtime.Provider.DestroySandbox(
+			executionContext.PodName,
+			executionContext.Namespace,
+			executionContext.Targets,
+			clusterConfig.Namespace,
+			clusterConfig.KeepNamespace,
+			clusterConfig.KeepNamespaceOnError,
+		)
+		if err != nil {
+			log.Errorf("Problem executing apb [%s] bind", executionContext.PodName)
+			e.finishWithError(err)
+			return
+		}
+		k8scli, err := clients.Kubernetes()
+		if err != nil {
+			log.Error("Something went wrong getting kubernetes client")
+			e.finishWithError(err)
+			return
+		}
+
+		if instance.Spec.Runtime >= 2 {
+			err := watchPod(executionContext.PodName, executionContext.Namespace,
+				k8scli.Client.CoreV1().Pods(executionContext.Namespace), e.updateDescription)
+			if err != nil {
+				log.Errorf("Bind action failed - %v", err)
+				e.finishWithError(err)
+				return
+			}
+		}
 
-	if instance.Spec.Runtime >= 2 {
-		err := watchPod(executionContext.PodName, executionContext.Namespace, k8scli.Client.CoreV1().Pods(executionContext.Namespace), stateUpdates)
+		creds, err := ExtractCredentials(
+			executionContext.PodName,
+			executionContext.Namespace,
+			instance.Spec.Runtime,
+		)
+		e.extractedCredentials = creds
```

Comment:
What if there is an err with `ExtractCredentials`? Do we want to set `e.extractedCredentials` before checking the err value? If this is a conscience decision then okay, just wanted to make sure we're not setting something invalid.

Reply:
It was actually originally written this way, I thought about it and the expectation is that creds will be nil if an error is returned from `ExtractCredentials`. Your comment is probably safer, however. Will update.